rfcs/text/typescript.md

```diff
@@ -0,0 +1,259 @@
+
+## Unit testing
+
+Converting to TypeScript means our spec files will also be TypeScript and therefore all code within will need to be typed correctly. While more effort than having JavaScript test files, this will help to weed out any bad code in our tests, leading to higher quality unit testing.
+
+There is also a guide in the [Jest documentation](https://jestjs.io/docs/getting-started#using-typescript) on how to set up Jest with TypeScript. We may want to add the [`@types/jest`](https://www.npmjs.com/package/@types/jest) module as described there.
+
+We may also need to add the [`@types/enzyme`](https://www.npmjs.com/package/@types/enzyme?activeTab=versions) module to add the type definitions for Enzyme.
+
+# Drawbacks
+
+## Conversion effort
+
+The first and main drawback is: it will likely be a large job to convert Carbon to TypeScript.
+
+Converting every single component, it's related styled components and spec files will take a while. Some may be simple, but some will be more complex, and running type checking against them may/will unearth bugs currently hidden. These will need to be fixed before the conversion will be complete. The flip side to that is we will likely find and fix bugs that are currently in our source code that we're not aware of.
+
+## Developer training
+
+Training will likely be needed for some, if not all, developers currently working on Carbon. Whilst training developers is never a bad thing, it will take time to complete and for developers to get comfortable using TypeScript on a day to day basis.
+
+## TypeScript lifespan
+
+TypeScript is generally expected to be around for the long run. However, if the JavaScript community moves away for it for whatever reason, future effort will be required to convert the library back to JavaScript.
+
+# Alternatives
+
+## Doing nothing
+
+The simplest alternative is simply to keep Carbon as it is currently. Write components in JavaScript and continue to provide separate TypeScript interfaces for them. This would be far less initial effort than conversion to TypeScript but long term we will continue to see issues raised and continue to have to maintain all of our component prop types in two places. This goes aginst the DRY principle of software development and will without a doubt continue to cause us problems.
+
+## Removing TypeScript support
+
+This is out of the question really as we know for a fact that a lot of consumers of Carbon are TypeScript projects. To be a good modern open source component library we should be supporting TypeScript.
+
+# Adoption strategy
+
+If this RFC PR is accepted we will need to create a task to carry out this conversion.
+
+Ideally the work will be carried out in a single PR (although it will be a large task!), with a commit per component, and once that is merged Carbon will be a TypeScript library.
```

Comment:
Could this work be done piecemeal and once we convert a component file to tsx do we have to do the associated spec files (including ones that have a given component composed within another) at the same time?

Reply:
If we could do this incrementally on a per component basis it would be much more preferable. Presumably the changes should be non breaking?

Reply:
@edleeks87 we don't have to but I think we should, we can have a Jira ticket per component to convert everything in that component folder to TypeScript. Everything that needs to be converted anyway

@chrisbarber86 yep it should be non breaking. I've updated the config detailed in the RFC to show how it can be set up to have js and ts components in parallel until all of them are converted (I've tested this in my test carbon-typescript repo as well). I've rewritten this bit to no longer recommend doing it in a single PR